// plus-editor.md/src/editor.js
import SimpleMDE from 'simplemde';
import 'simplemde/dist/simplemde.min.css';

import md from './tools/md';
import addImage from './plugin/addImg';
let toolbar = [{
    name: "bold",
    className: "fa fa-bold",
    action: SimpleMDE.toggleBold,
    title: "粗体"
}, {
    name: "italic",
    className: "fa fa-italic",
    action: SimpleMDE.toggleItalic,
    title: "斜体"
}, {
    name: "strikethrough",
    className: "fa fa-strikethrough",
    action: SimpleMDE.toggleStrikethrough,
    title: "删除线"
}, {
    name: "horizontal-rule",
    className: "fa fa-minus",
    action: SimpleMDE.drawHorizontalRule,
    title: "水平线"
}, '|', {
    name: "heading-1",
    className: "fa fa-header fa-header-x fa-header-1",
    action: SimpleMDE.toggleHeading1,
    title: "H1"
}, {
    name: "heading-2",
    className: "fa fa-header fa-header-x fa-header-2",
    action: SimpleMDE.toggleHeading2,
    title: "H2"
}, {
    name: "heading-3",
    className: "fa fa-header fa-header-x fa-header-3",
    action: SimpleMDE.toggleHeading3,
    title: "H3"
}, '|', {
    name: "code",
    className: "fa fa-code",
    action: SimpleMDE.toggleCodeBlock,
    title: "代码"
}, {
    name: "quote",
    className: "fa fa-quote-left",
    action: SimpleMDE.toggleBlockquote,
    title: "引用"
}, {
    name: "unordered-list",
    className: "fa fa-list-ul",
    action: SimpleMDE.toggleUnorderedList,
    title: "无序列表"
}, {
    name: "ordered-list",
    className: "fa fa-list-ol",
    action: SimpleMDE.toggleOrderedList,
    title: "有序列表"
}, '|', {
    name: "link",
    className: "fa fa-link",
    action: SimpleMDE.drawLink,
    title: "超链接"
}, {
    name: "image",
    className: "fa fa-picture-o",
    action: SimpleMDE.drawImage,
    title: "外链图片"
}, {
    name: 'addImg',
    className: "fa fa-plus",
    action: addImage,
    title: "上传图片"
}, {
    name: "table",
    className: "fa fa-table",
    action: SimpleMDE.drawTable,
    title: "插入表格"
}, '|', {
    name: "preview",
    className: "fa fa-eye no-disable",
    action: SimpleMDE.togglePreview,
    title: "预览"
}, {
    name: "side-by-side",
    className: "fa fa-columns no-disable no-mobile",
    action: SimpleMDE.toggleSideBySide,
    title: "实时预览"
}, {
    name: "fullscreen",
    className: "fa fa-arrows-alt no-disable no-mobile",
    action: SimpleMDE.toggleFullScreen,
    title: "全屏预览"
}, 'guide'];
const placeholder = "开始你的表演...";
const __default = {
    toolbar: toolbar,
    placeholder,
    autoDownloadFontAwesome: false,
    renderingConfig: {
        codeSyntaxHighlighting: true
    },
    insertTexts: {
        addImg: ["@![图片描述](", "#ID#)"]
    },
    fileApiPath: '/api/v2/files',
    uploadFile(file, cb) {
        console.log(file);
        cb(1);
    }
};

export default class PlusEditor {
    constructor(options) {
        if(!options.element) {
            throw new Error("element参数为必须项!");
        }
        this.options = Object.assign({}, __default, options);
        this.options.previewRender = md(this.options);
        return new SimpleMDE(this.options);
    }
}
