// plus-editor.md/webpack.config.babel.js
import path from 'path';
import webpack from 'webpack';
import CleanWebpackPlugin from 'clean-webpack-plugin';
import ExtractTextPlugin from 'extract-text-webpack-plugin';

// 引入 baseConfig
import baseConfig, { options } from './build/base.config';

/**
 * 定义输出路径
 * @type {String}
 */
const PublicPath = path.resolve(__dirname, "src"),
    BuildRoot = path.resolve(PublicPath, "../dist");

const CONFIG = {
    /**
     * 定义入口文件
     * @type {Object}
     */
    entry: {
        pluseditor: './src/index.js',
    },

    /**
     * 定义输出
     * @type {Object}
     */
    output: {

        /**
         * 输出路径
         * @type {String}
         * 必须为绝对路径
         */
        path: BuildRoot,

        /**
         * 文件名
         * @type {String}
         * 开发环境不压缩      -> xx.js
         * 生产环境启用压缩    -> xx.min.js
         */
        filename: options.optimizeMinimize ? '[name].min.js' : '[name].js',

        /**
         * library
         * @type {String}
         */
        library: 'PlusEditor',
        libraryTarget: 'umd',
    },

    module: {
        loaders: [
            { test: /\.js/, loader: "babel-loader", exclude: /node_modules/ },
            { test: /\.css$/, loader: ExtractTextPlugin.extract({ fallback: "style-loader", use: "css-loader" }) },
        ],
    },

    plugins: [
        new CleanWebpackPlugin('./*', {
            root: BuildRoot,
            verbose: true,
            dry: false
        }),

        new ExtractTextPlugin("pluseditor.css"),
    ],

    externals: [],
};


export default Object.assign(baseConfig, CONFIG);
