// plus-editor.md/build/base.config.js
/**
 * webpack基础配置
 *
 * @author josnleex <jsonleex@163.com> (17-09-23 05:11)
 * 
 */

import webpack from "webpack";

/**
 * 命令行工具，用于获取命令行参数
 */
import yargs from 'yargs';

/**
 * 提取命令行参数配置
 * 
 * - p      -> 环境配置 是否优化和压缩代码
 * 
 */
export const options = yargs
    .alias('p', 'optimize-minimize')
    .alias('d', 'debug')
    .argv;

const baseConfig = {
    /**
     * 入口配置
     * @type { string | object | array }
     *
     * entry: "./app/entry",
     * 
     * entry: ["./app/entry1", "./app/entry2"],
     *
     * entry: {
     *     a: "./app/entry-a",
     *     b: ["./app/entry-b1", "./app/entry-b2"]
     * },
     * 
     */
    entry: undefined,

    /**
     * 输出配置 (path filename publicPath)
     *
     *  path        -> 保存路径
     *
     *  filename    -> 文件名
     *
     *  publicPath  -> 输出目录相对于HTML页面的url (分布式 CDN 情况下非常实用)
     *  
     */
    output: undefined,

    /**
     * [externals description]
     * @type {[type]}
     */
    externals: undefined,

    /**
     * 模块配置
     * 
     */
    module: {
        /**
         * 加载器 指定 xx 文件由 xx-loader 解析
         * 
         * @type {Array}
         *
         * 基本格式：{ 
         *     test: REG,                   // 解析文件类型
         *     include：/src/,              // 解析指定目录
         *     exclude: /node_modules/,     // 不用解析的文件目录
         *     loader: xx-loader,           // 指定加载器
         *     option: {},                  // xx-loader 的其他配置项
         * }
         */
        loaders: [
            { test: /\.js/, loader: "babel-loader", exclude: /node_modules/ }
        ]
    },

    /**
     * 为浏览器devtools添加元信息来增强调试
     * @type {String}
     *
     * 使用 source-map 会降低构建速度
     * 
     */
    // leex: 此处修改为根据环境变量配置 baseConfig.devtool
    // devtool: "source-map", 

    /**
     * 插件
     * @type {Array}
     * 
     */
    plugins: [
        /**
         *
         * DefinePlugin (直译：webpack的定义插件 --此处要被怼 😂)
         * 
         * 官方文档 -> https://webpack.js.org/plugins/define-plugin/
         *
         */
        new webpack.DefinePlugin({
            'process.env': {
                NODE_ENV: JSON.stringify(options.optimizeMinimize ? 'production' : 'development')
            }
        })
    ]
};

if(options.optimizeMinimize) {
    baseConfig.devtool = 'source-map';
}

export default baseConfig;
